File: scripts/bindings/ggplot_palette_functions.py
"""FIX-GG04: discrete palette values, lookup errors and immutable host publication."""
import copy,json,math,sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
registry=c.ExtensionRegistry.example();records=[]
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(640,360).dpi(96).basis('current')
cases=[t for t in json.loads((ROOT/'fixtures/parity/ggplot2/palette-functions.json').read_text())['cases'] if t['family']=='discrete']
cases+=json.loads((ROOT/'fixtures/parity/ggplot2/discrete-palette-lookup.json').read_text())['cases']
cases+=json.loads((ROOT/'fixtures/parity/ggplot2/discrete-palette-missing-paint.json').read_text())['cases']
def key(v):return 'Null' if v is None else {'Text':v}
def descriptor(t):
 s={'missing_paint_is_na':t['channel']=='colour','training':'Eligible','function':{'Ordinal':{'domain':[],'range':[],'unknown':{'Explicit':None}}},'ggplot':{'Discrete':{'limits':list(map(key,['a','b','c','d'])) if t['limits']=='full' else None,'levels':None,'drop':True,'na_translate':True,'palette':{'Hue':{'h':[15,375],'chroma':100,'luminance':65,'start':0,'reverse':False}}}},'palette_function':{'operation':{'id':t.get('operation','example.scale_palette'),'version':'1'},'parameters':{'mode':t['mode'],'channel':t['channel']}}}
 if t.get('guide_mode')=='hidden':s['guide']='Hidden'
 if t.get('guide_mode')=='first':s['guide']={'Discrete':{'breaks':[key('a')]}}
 return s
def layer(t):return c.points().name('marks') if t['channel']=='colour' else c.points().name('marks').numeric_scale('Size' if t['channel']=='size' else 'Alpha','v',descriptor(t))
def data_for(t):return c.Data.columns({'x':c.column(list(map(float,range(len(t['inputs'])))),kind='float64'),'v':c.column(t['inputs'],kind='string').nullable(True)})
def build(t,d):
 draft=c.plot(d).with_registry(registry).profile('Ggplot2_4_0_3')
 if t['channel']=='colour':
  scale=c.color_mapped('v',descriptor(t))
  if t.get('na_mode','NA')!='NA':scale=scale.missing(paint('#00ff00' if t['na_mode']=='green' else '#00000000'))
  return draft.aes(c.aes().x('x').y(1.).color('v').color_scale('v')).scale(scale).layer(layer(t)).build()
 return draft.aes(c.aes().x('x').y(1.)).layer(layer(t)).build()
def compare(a,b):
 if isinstance(a,(int,float)) and isinstance(b,(int,float)):assert math.isclose(a,b,rel_tol=0,abs_tol=5e-14),(a,b)
 elif isinstance(a,list) and isinstance(b,list):
  assert len(a)==len(b),(a,b)
  for x,y in zip(a,b):compare(x,y)
 elif isinstance(a,dict) and isinstance(b,dict):
  assert a.keys()==b.keys(),(a,b)
  for k,v in a.items():compare(v,b[k])
 else:assert a==b,(a,b)
def paint(v):return dict(zip(('red','green','blue','alpha'),[0,0,0,0] if v is None else [int(v[i:i+2],16) for i in (1,3,5)]+[int(v[7:9],16) if len(v)==9 else 255]))
def check(t,chart):
 state=chart.semantics()['layers'][0]
 if t['channel']=='colour':
  legend=state.get('color_legend')
  compare([s['color'] for s in state.get('styles',[])],[paint(v) for v in t['result']['mapped'] if v is not None])
  if legend is None:
   assert not t['result'].get('keys'),t
   return {'mapped':None,'domain':None,'guide':[],'styles':state.get('styles',[])}
  spec=legend['mapping'];entries=legend['entries']
 else:
  spec=state['numeric_scales']['Size' if t['channel']=='size' else 'Alpha']['scale'];entries=[e for g in state.get('discrete_value_guides',{}).values() for e in g]
 ordinal=spec['function']['Ordinal'];domain=[k.get('Text') if isinstance(k,dict) else None for k in ordinal['domain']]
 def fallback():
  if t['channel']!='colour' or spec.get('missing_paint_is_na',False):return None
  missing=legend['missing'];return '#'+''.join(f'{missing[k]:02x}' for k in ('red','green','blue'))+(f"{missing['alpha']:02x}" if missing['alpha']!=255 else '')
 def sample(v):
  if v is None or v not in domain:return fallback()
  index=domain.index(v);raw=ordinal['range'][index]
  if raw['kind'] in ('Number','Text'):return raw.get('value')
  return fallback() if index in spec.get('palette_fallback_indices',[]) else None
 actual=[sample(v) for v in t['inputs']];compare(actual,t['result']['mapped'])
 expected=next(iter(t['result'].get('keys',[])),{'values':[],'labels':[],'mapped':[]})
 if t['channel']=='colour':
  wanted=[['NA' if label is None else label,paint(value)] for label,value in zip(expected['labels'],expected['mapped'])];compare(entries,wanted)
 else:
  actual_keys={'values':[e['key'].get('Text') if isinstance(e['key'],dict) else None for e in entries],'labels':[e['label'] for e in entries],'mapped':[sample(e['key'].get('Text') if isinstance(e['key'],dict) else None) for e in entries]};compare(actual_keys,expected)
 return {'mapped':actual,'domain':domain,'guide':entries,'styles':state.get('styles',[])}
for index,t in enumerate(cases):
 owned=[]
 try:
  d=data_for(t);owned.append(d);p=build(t,d);owned.append(p);wire=p.to_json();assert json.loads(wire)['version']==37
  restored=c.Plot.from_json(wire,registry);owned.append(restored);assert restored.to_json()==wire
  for state in ('original','edited'):
   current=restored if state=='original' else restored.edit().layer('marks',layer(t)).build()
   if current is not restored:owned.append(current)
   chart=current.chart();owned.append(chart);actual=check(t,chart);assert 'error' not in t['result'],t
   records.append({'index':index,'state':state,**actual})
  sample=None
  if not t.get('guide_mode') and not t.get('na_mode') and t['limits']=='none' and t['mode']=='named' and t['population'] in ('ordinary','missing'):sample=f"{t['channel']}-{t['population']}"
  if t.get('na_mode') and t['mode']=='missing' and t['population']=='missing':sample='na-'+t['na_mode'].lower()
  if sample:
   request=output.request(restored,options);owned.append(request);frame=request.prepare();owned.append(frame)
   for fmt in ('svg','pdf','png'):(out/f'{sample}.{fmt}').write_bytes(frame.export(fmt))
 except c.ChartError as error:
  assert 'error' in t['result'],(t,str(error));assert error.code=='CHART_VALIDATION',(t,error.code)
  records.append({'index':index,'error':error.code})
 finally:
  for obj in reversed(owned):obj.dispose()
for channel in ('colour','size','alpha'):
 for mode in ('full','named'):
  selected={t['population']:t for t in cases if not t.get('guide_mode') and not t.get('na_mode') and t['channel']==channel and t['limits']=='none' and t['mode']==mode};owned=[]
  try:
   original=data_for(selected['ordinary']);owned.append(original);p=build(selected['ordinary'],original);owned.append(p);chart=p.chart();owned.append(chart);wire=p.to_json();request=output.request(p,options);owned.append(request);held=request.prepare();owned.append(held);scene=held.scene()
   for population in ('missing','all_missing','empty','ordinary'):
    t=selected[population];replacement=data_for(t);owned.append(replacement);tx=chart.transaction();owned.append(tx);builder=tx.replace(original,replacement);owned.append(builder);update=builder.build();owned.append(update);assert 'Applied' in chart.commit(update)
    fresh=build(t,replacement);owned.append(fresh);batch=fresh.chart();owned.append(batch);actual=check(t,chart);assert actual==check(t,batch);assert p.to_json()==wire and held.scene()==scene
    records.append({'channel':channel,'mode':mode,'replacement':population,**actual})
  finally:
   for obj in reversed(owned):obj.dispose()
for variant in ('missing_registration','native_only','invalid_parameters','downgrade'):
 t=copy.deepcopy(cases[0]);owned=[]
 if variant=='missing_registration':t['operation']='example.missing_palette'
 if variant=='native_only':t['operation']='example.native_scale_palette'
 if variant=='invalid_parameters':t['mode']='invalid'
 try:
  d=data_for(t);owned.append(d);p=build(t,d);owned.append(p);wire=p.to_json()
  if variant=='downgrade':wire=json.loads(wire);wire['version']=36;restored=c.Plot.from_json(json.dumps(wire),registry);owned.append(restored)
  raise AssertionError('expected rejection: '+variant)
 except c.ChartError as error:records.append({'rejection':variant,'code':error.code})
 finally:
  for obj in reversed(owned):obj.dispose()
expected=sum(1 if 'error' in t['result'] else 2 for t in cases)+28
assert len(records)==expected,(len(records),expected)
(out/'records.json').write_text(json.dumps(records,indent=2));output.dispose();registry.dispose()
print(f'PASS Python: {len(records)} discrete palette states including 24 replacements and four rejection contracts.')

File: crates/chart-core/tests/ggplot_palette_functions.rs
//! FIX-GG04: count palettes against independently captured public constructor builds.
use chart_core::{
    ChartResult, Revision,
    grammar::*,
    interpolate::{Number, Value},
    scales::*,
};
use serde_json::{Value as Json, json};
use std::sync::{Arc, Mutex};
struct Palette(Arc<Mutex<Vec<usize>>>);
impl CustomScalePalette for Palette {
    fn descriptor(&self) -> ExtensionDescriptor {
        ExtensionDescriptor::batch("test.palette", Revision::new(1), true)
    }
    fn validate(&self, _: &Json) -> ChartResult<()> {
        Ok(())
    }
    fn evaluate(&self, input: ScalePaletteInput<'_>) -> ChartResult<ScalePaletteOutput> {
        let ScalePaletteDomain::Count(count) = input.domain else {
            panic!("discrete input expected")
        };
        self.0.lock().unwrap().push(count);
        let mode = input.parameters["mode"].as_str().unwrap();
        let channel = input.parameters["channel"].as_str().unwrap();
        let mut values = (1..=count)
            .map(|i| {
                let t = i as f64 / count.max(1) as f64;
                match channel {
                    "colour" => Value::Text(if t < 0.5 { "#ff0000" } else { "#0000ff" }.into()),
                    "size" => Value::Number(Number(1. + 4. * t)),
                    _ => Value::Number(Number(t)),
                }
            })
            .collect::<Vec<_>>();
        match mode {
            "short" => values.truncate(1),
            "empty" => values.clear(),
            "missing" if values.len() > 1 => values[1] = Value::Missing,
            _ => {}
        }
        let names = (mode == "named").then(|| {
            (0..values.len())
                .map(|i| ["c", "b", "a", "d"][i % 4].to_owned())
                .collect()
        });
        Ok(ScalePaletteOutput {
            values: (mode != "null").then_some(values),
            names,
        })
    }
}
fn spec(c: &Json) -> MappedScaleSpec {
    let mut result = MappedScaleSpec::authored(ScaleFunctionSpec::Ordinal(OrdinalSpec::default()))
        .with_ggplot(GgplotScalePolicy::Discrete {
            empty_population: false,
            limits: (c["limits"] == "full").then(|| {
                ["a", "b", "c", "d"]
                    .map(|s| ScaleKey::Text(s.into()))
                    .to_vec()
            }),
            levels: None,
            drop: true,
            na_translate: true,
            palette: GgplotDiscretePalette::Hue(Default::default()),
        })
        .unwrap()
        .with_palette_function(ScalePaletteOperation {
            operation: OperationRef {
                id: "test.palette".into(),
                version: Revision::new(1),
            },
            parameters: json!({"channel":c["channel"],"mode":c["mode"]}),
        })
        .unwrap();
    if c["guide_mode"] == "hidden" {
        result.guide = Some(Box::new(GgplotScaleGuide::Hidden));
    }
    if c["guide_mode"] == "first" {
        result.guide = Some(Box::new(GgplotScaleGuide::Discrete(GgplotDiscreteGuide {
            breaks: Some(vec![ScaleKey::Text("a".into())]),
            ..Default::default()
        })));
    }
    result.missing_paint_is_na = c["channel"] == "colour";
    result
}
fn key(j: &Json) -> ScaleKey {
    j.as_str()
        .map_or(ScaleKey::Null, |s| ScaleKey::Text(s.into()))
}
fn encode(v: Value) -> Json {
    match v {
        Value::Missing => Json::Null,
        Value::Text(s) => json!(s),
        Value::Number(n) => json!(n.0),
        _ => panic!("palette primitive"),
    }
}
#[test]
fn discrete_palette_functions_match_180_public_constructor_cases() {
    assert_cases(
        include_str!("../../../fixtures/parity/ggplot2/palette-functions.json"),
        180,
    );
}
#[test]
fn discrete_palette_lookup_matches_360_hidden_and_restricted_guide_cases() {
    assert_cases(
        include_str!("../../../fixtures/parity/ggplot2/discrete-palette-lookup.json"),
        360,
    );
}
fn assert_cases(source: &str, expected_count: usize) {
    let fixture: Json = serde_json::from_str(source).unwrap();
    let mut count = 0;
    for c in fixture["cases"]
        .as_array()
        .unwrap()
        .iter()
        .filter(|c| c["family"] == "discrete")
    {
        count += 1;
        let calls = Arc::new(Mutex::new(vec![]));
        let mut registry = ExtensionRegistry::new();
        registry
            .register_scale_palette(Arc::new(Palette(calls.clone())))
            .unwrap();
        let authored = spec(c);
        let inputs = c["inputs"]
            .as_array()
            .unwrap()
            .iter()
            .map(key)
            .collect::<Vec<_>>();
        MappedScale::new_with_registry(authored.clone(), &registry).unwrap();
        assert!(calls.lock().unwrap().is_empty());
        let result = (|| -> ChartResult<Json> {
            let trained = authored.trained_keys_with_registry(&inputs, &registry)?;
            let mapping = MappedScale::new_with_registry(trained, &registry)?;
            let mapped = inputs
                .iter()
                .map(|k| mapping.category(Some(k)).map(encode))
                .collect::<ChartResult<Vec<_>>>()?;
            let entries = mapping.discrete_guide_entries()?.unwrap_or_default();
            let mut values = vec![];
            let mut labels = vec![];
            let mut colors = vec![];
            for e in entries {
                values.push(match &e.key {
                    ScaleKey::Null => Json::Null,
                    ScaleKey::Text(s) => json!(s),
                    _ => panic!("text key"),
                });
                labels.push(e.label);
                colors.push(encode(mapping.category(Some(&e.key))?));
            }
            Ok(
                json!({"mapped":mapped,"keys":if values.is_empty(){json!([])}else{json!([{"values":values,"labels":labels,"mapped":colors}])}}),
            )
        })();
        if c["result"].get("error").is_some() {
            assert!(result.is_err(), "{c}: unexpected preparation outcome");
        } else {
            compare(&result.unwrap(), &c["result"]);
        }
        let expected = c["calls"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v["values"][0].as_u64().unwrap() as usize)
            .collect::<Vec<_>>();
        assert_eq!(*calls.lock().unwrap(), expected, "{c}");
    }
    assert_eq!(count, expected_count);
}

fn compare(actual: &Json, expected: &Json) {
    match (actual, expected) {
        (Json::Number(a), Json::Number(b)) => assert!(
            (a.as_f64().unwrap() - b.as_f64().unwrap()).abs() < 5e-14,
            "{actual} != {expected}"
        ),
        (Json::Array(a), Json::Array(b)) => {
            assert_eq!(a.len(), b.len());
            for (a, b) in a.iter().zip(b) {
                compare(a, b);
            }
        }
        (Json::Object(a), Json::Object(b)) => {
            assert_eq!(a.len(), b.len());
            for (k, v) in a {
                compare(v, &b[k]);
            }
        }
        _ => assert_eq!(actual, expected),
    }
}

#[test]
fn primary_palette_roundtrip_prepares_all_540_reference_outcomes() {
    use chart_core::prelude::*;
    for source in [
        include_str!("../../../fixtures/parity/ggplot2/palette-functions.json"),
        include_str!("../../../fixtures/parity/ggplot2/discrete-palette-lookup.json"),
    ] {
        let fixture: Json = serde_json::from_str(source).unwrap();
        for c in fixture["cases"]
            .as_array()
            .unwrap()
            .iter()
            .filter(|c| c["family"] == "discrete")
        {
            let calls = Arc::new(Mutex::new(vec![]));
            let mut registry = ExtensionRegistry::new();
            registry
                .register_scale_palette(Arc::new(Palette(calls.clone())))
                .unwrap();
            let registry = Arc::new(registry);
            let inputs = c["inputs"].as_array().unwrap();
            let data = Data::columns()
                .column("x", (0..inputs.len()).map(|i| i as f64).collect::<Vec<_>>())
                .column(
                    "v",
                    inputs
                        .iter()
                        .map(|v| v.as_str().map(String::from))
                        .collect::<Vec<_>>(),
                )
                .build()
                .unwrap();
            let draft = plot(data)
                .profile(Profile::Ggplot2_4_0_3)
                .extensions(registry.clone());
            let p = if c["channel"] == "colour" {
                draft
                    .aes(aes().x("x").y(1.).color("v").color_scale("v"))
                    .scale(color_mapped("v", spec(c)))
                    .layer(points())
                    .build()
                    .unwrap()
            } else {
                draft
                    .aes(aes().x("x").y(1.))
                    .layer(points().numeric_scale(
                        if c["channel"] == "size" {
                            NumericAesthetic::Size
                        } else {
                            NumericAesthetic::Alpha
                        },
                        "v",
                        spec(c),
                    ))
                    .build()
                    .unwrap()
            };
            let wire = p.to_json().unwrap();
            assert_eq!(serde_json::from_str::<Json>(&wire).unwrap()["version"], 37);
            assert!(calls.lock().unwrap().is_empty());
            assert!(Plot::from_json(&wire).is_err());
            let restored = Plot::from_json_with_extensions(&wire, registry).unwrap();
            assert_eq!(restored.to_json().unwrap(), wire);
            let result = restored.chart().unwrap().prepare();
            if c["channel"] == "colour"
                && let Ok(prepared) = &result
            {
                let actual = prepared.layers()[0]
                    .marks()
                    .iter()
                    .map(|m| m.style.color)
                    .collect::<Vec<_>>();
                let expected = c["result"]["mapped"]
                    .as_array()
                    .unwrap()
                    .iter()
                    .filter_map(Json::as_str)
                    .map(|v| chart_core::color::parse_r(v).unwrap().resolve())
                    .collect::<Vec<_>>();
                assert_eq!(actual, expected, "{c}");
            }
            assert_eq!(
                result.is_err(),
                c["result"].get("error").is_some(),
                "{c}: unexpected preparation outcome"
            );
        }
    }
}

#[test]
fn palette_na_elements_and_missing_input_replacements_match_54_source_draws() {
    use chart_core::prelude::*;
    let fixture: Json = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/discrete-palette-missing-paint.json"
    ))
    .unwrap();
    let cases = fixture["cases"].as_array().unwrap();
    assert_eq!(cases.len(), 54);
    for c in cases {
        let mut registry = ExtensionRegistry::new();
        registry
            .register_scale_palette(Arc::new(Palette(Arc::new(Mutex::new(vec![])))))
            .unwrap();
        let inputs = c["inputs"].as_array().unwrap();
        let data = Data::columns()
            .column("x", (0..inputs.len()).map(|i| i as f64).collect::<Vec<_>>())
            .column(
                "v",
                inputs
                    .iter()
                    .map(|v| v.as_str().map(String::from))
                    .collect::<Vec<_>>(),
            )
            .build()
            .unwrap();
        let mut scale = color_mapped("v", spec(c));
        if c["na_mode"] != "NA" {
            scale = scale.missing(
                chart_core::color::parse_r(if c["na_mode"] == "green" {
                    "#00ff00"
                } else {
                    "#00000000"
                })
                .unwrap(),
            );
        }
        let p = plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .extensions(Arc::new(registry))
            .aes(aes().x("x").y(1.).color("v").color_scale("v"))
            .scale(scale)
            .layer(points())
            .build()
            .unwrap();
        let result = p.chart().unwrap().prepare();
        if c["result"].get("error").is_some() {
            assert!(result.is_err(), "{c}");
            continue;
        }
        let prepared = result.unwrap();
        let layer = &prepared.layers()[0];
        let marks = layer.marks();
        assert_eq!(
            marks.len(),
            c["result"]["point_count"].as_u64().unwrap() as usize,
            "{c}"
        );
        let expected = c["result"]["mapped"]
            .as_array()
            .unwrap()
            .iter()
            .filter_map(Json::as_str)
            .map(|s| chart_core::color::parse_r(s).unwrap().resolve())
            .collect::<Vec<_>>();
        assert_eq!(
            marks.iter().map(|m| m.style.color).collect::<Vec<_>>(),
            expected,
            "{c}"
        );
        let entries = &layer.color_legend().unwrap().entries;
        let expected = c["result"]["keys"][0]["mapped"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| {
                v.as_str().map_or(
                    chart_core::scene::Color {
                        red: 0,
                        green: 0,
                        blue: 0,
                        alpha: 0,
                    },
                    |v| chart_core::color::parse_r(v).unwrap().resolve(),
                )
            })
            .collect::<Vec<_>>();
        assert_eq!(
            entries.iter().map(|e| e.1).collect::<Vec<_>>(),
            expected,
            "{c}"
        );
    }
}

#[test]
fn malformed_palette_results_and_fallback_metadata_are_rejected() {
    struct Invalid(bool);
    impl CustomScalePalette for Invalid {
        fn descriptor(&self) -> ExtensionDescriptor {
            ExtensionDescriptor::batch("test.palette", Revision::new(1), true)
        }
        fn validate(&self, _: &Json) -> ChartResult<()> {
            Ok(())
        }
        fn evaluate(&self, _: ScalePaletteInput<'_>) -> ChartResult<ScalePaletteOutput> {
            Ok(if self.0 {
                ScalePaletteOutput {
                    values: Some(vec![Value::Missing]),
                    names: Some(vec![]),
                }
            } else {
                ScalePaletteOutput {
                    values: Some(vec![
                        Value::Missing;
                        chart_core::interpolate::MAX_VALUES + 1
                    ]),
                    names: None,
                }
            })
        }
    }
    let source = spec(&json!({"limits":"none","channel":"colour","mode":"full"}));
    for malformed_names in [true, false] {
        let mut registry = ExtensionRegistry::new();
        registry
            .register_scale_palette(Arc::new(Invalid(malformed_names)))
            .unwrap();
        let error = source
            .trained_keys_with_registry(&[ScaleKey::Text("a".into())], &registry)
            .unwrap_err();
        assert_eq!(
            error.code,
            if malformed_names {
                chart_core::DiagnosticCode::Validation
            } else {
                chart_core::DiagnosticCode::ResourceLimit
            }
        );
    }
    let mut registry = ExtensionRegistry::new();
    registry
        .register_scale_palette(Arc::new(Palette(Arc::new(Mutex::new(vec![])))))
        .unwrap();
    let mut forged = source
        .trained_keys_with_registry(&[ScaleKey::Text("a".into())], &registry)
        .unwrap();
    forged.palette_fallback_indices = vec![1];
    assert!(MappedScale::new_with_registry(forged, &registry).is_err());
}
